### PR TITLE
Asynchronous CVR Import

### PR DESCRIPTION
**Pull request by dmzimmerman**
This PR includes changes to support asynchronous CVR import, namely:

- additional states and events (including some renamed events) in the ASM; the new states, `CVRS_IMPORTING` and `BALLOT_MANIFESTS_OK_AND_CVRS_IMPORTING`, are the analogues to `CVRS_OK` and `BALLOT_MANIFESTS_AND_CVRS_OK` that indicate that a CVR import is in progress
- a complete reorganization of the `CVRExportImport` endpoint that returns (almost) immediately to the client and uses a background thread to do the import and update the system state
- a modification of transaction retries to use a more traditional exponential backoff algorithm

This PR is based on the `misc-server-fixes` branch, and should _not_ be merged until that branch is, and until any necessary client changes are made.

### CODE REVIEW

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java [hunk 1]
```diff
@@ -101,333 +94,573 @@ public String endpointName() {
...
+          the_file.setStatus(FileStatus.IMPORTED_AS_CVR_EXPORT);
+          Persistence.saveOrUpdate(the_file);
+        } else {
           try {
-            Persistence.rollbackTransaction();
-          } catch (final PersistenceException ex) {
-            // not much we can do about it
+            cleanup(the_file.county(), true);
+          } catch (final PersistenceException e) {
+            // if we couldn't clean up, there's not much we can do about it
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java]: Do we want to log these cases?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java [hunk 2]
```diff
@@ -101,333 +94,573 @@ public String endpointName() {
...
       } else {
-        badDataContents(the_response, "attempt to import a file without a verified hash");
-      }
-    } catch (final JsonParseException e) {
-      badDataContents(the_response, "malformed request: " + e.getMessage());
-    } finally {
-      // signal that we're done with the import
-      synchronized (COUNTIES_RUNNING) {
-        COUNTIES_RUNNING.remove(county.id());
+        final long max_delay_factor = Math.min(1, Math.round(exponentiated));
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java]: Should this be `Math.max()`?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java [hunk 3]
```diff
@@ -406,6 +409,66 @@ private void checkForFlush() {
...
+          Persistence.commitTransaction();
+          success = true;
+        }
+      } catch (final PersistenceException e) {
+        // something went wrong, let's try again
+        if (Persistence.canTransactionRollback()) {
+          try {
+            Persistence.rollbackTransaction();
+          } catch (final PersistenceException ex) {
+            // not much we can do about it
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java]: Log here?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java [hunk 4]
```diff
@@ -406,6 +409,66 @@ private void checkForFlush() {
...
+      }
+    }
+    // we always need a running transaction
+    Persistence.beginTransaction();
+    if (success && retries > 1) {
+      Main.LOGGER.info("updated state machine for county " + my_county.id() + 
+                       " in " + retries + " tries");
+    } else if (!success) {
+      throw new PersistenceException("could not update state machine for county " + 
+                                     my_county.id() + " after " + (retries - 1) + " tries");
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java]: Should we really subtract 1 from `retries` here?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/util/ExponentialBackoffHelper.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+   * @param the_retries The number of retries so far.
+   */
+  public static long exponentialBackoff(final int the_retries, final long the_unit_delay) {
+    final double exponentiated = Math.pow(2,  the_retries);
+    long multiplier = 1;
+    
+    if (Double.isNaN(exponentiated)) {
+      multiplier = 1;
+    } else {
+      final long max_delay_factor = Math.min(1, Math.round(exponentiated));
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/util/ExponentialBackoffHelper.java]: As before, should this be `Math.max()`?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/util/ExponentialBackoffHelper.java [hunk 6]
```diff
@@ -41,7 +41,7 @@ public static long exponentialBackoff(final int the_retries, final long the_unit
     if (Double.isNaN(exponentiated)) {
       multiplier = 1;
     } else {
-      final long max_delay_factor = Math.min(1, Math.round(exponentiated));
+      final long max_delay_factor = Math.max(1, Math.round(exponentiated));
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/util/ExponentialBackoffHelper.java]: Cool, you caught this.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java [hunk 7]
```diff
@@ -206,7 +206,7 @@ public String endpointBody(final Request the_request, final Response the_respons
     /**
      * The number of milliseconds to sleep between transaction retries.
      */
-    private static final long TRANSACTION_SLEEP_MSEC = 1000;
+    private static final long TRANSACTION_SLEEP_MSEC = 10;
```
Comment (ranweiler) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java]: :+1:

File: server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java (hunk 4 above)
Comment (dmzimmerman) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java]: No. I thought I’d caught all of those. I will fix that before merging.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java [hunk 8]
```diff
@@ -101,333 +94,573 @@ public String endpointName() {
...
-          response.put("records_deleted", deleted);
+    }
+    
+    try {
+      final UploadedFile file =
+          Main.GSON.fromJson(the_request.body(), UploadedFile.class);
+      if (file == null) {
+        badDataContents(the_response, "nonexistent file");
+      } else if (!file.county().id().equals(county.id())) {
+        unauthorized(the_response, "county " + county.id() + " attempted to import " +
```
[the remark this answers is not in the document]
Comment (dmzimmerman) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java]: Correct

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java [hunk 9]
```diff
@@ -101,333 +94,573 @@ public String endpointName() {
...
-
-        for (final CountyContestResult ccr : contest_results) {
-          if (ccr.minMargin() == 0) { 
-            // this is a tied contest
-            final ContestToAudit cta = new ContestToAudit(ccr.contest(), 
-                                                          AuditReason.TIED_CONTEST, 
-                                                          AuditType.NOT_AUDITABLE);
-            dosdb.updateContestToAudit(cta);
-            result = result + 1;
+        final InputStreamReader bmi_isr = new InputStreamReader(ufs.inputStream(), "UTF-8");
```
[the remark this answers is not in the document]
Comment (dmzimmerman) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java]: Probably. I can fix it.

**Pull request merged**